sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -90,21 +94,74 @@ trait BaseLimitExec extends UnaryExecNode with CodegenSupport {
-
-  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
-
-  override def outputPartitioning: Partitioning = child.outputPartitioning
-}
+case class GlobalLimitExec(limit: Int, child: SparkPlan) extends UnaryExecNode {
+  override def output: Seq[Attribute] = child.output
 
-/**
- * Take the first `limit` elements of the child's single output partition.
- */
-case class GlobalLimitExec(limit: Int, child: SparkPlan) extends BaseLimitExec {
+  protected override def doExecute(): RDD[InternalRow] = {
+    // This logic is mainly copyed from `SparkPlan.executeTake`.
+    // TODO: combine this with `SparkPlan.executeTake`, if possible.
+    val childRDD = child.execute()
+    val totalParts = childRDD.partitions.length
+    var partsScanned = 0
+    var totalNum = 0
+    var resultRDD: RDD[InternalRow] = null
+    while (totalNum < limit && partsScanned < totalParts) {
+      // The number of partitions to try in this iteration. It is ok for this number to be
+      // greater than totalParts because we actually cap it at totalParts in runJob.
+      var numPartsToTry = 1L
+      if (partsScanned > 0) {
+        // If we didn't find any rows after the previous iteration, quadruple and retry.
+        // Otherwise, interpolate the number of partitions we need to try, but overestimate
+        // it by 50%. We also cap the estimation in the end.
+        val limitScaleUpFactor = Math.max(sqlContext.conf.limitScaleUpFactor, 2)
+        if (totalNum == 0) {
+          numPartsToTry = partsScanned * limitScaleUpFactor
+        } else {
+          // the left side of max is >=1 whenever partsScanned >= 2
+          numPartsToTry = Math.max((1.5 * limit * partsScanned / totalNum).toInt - partsScanned, 1)
+          numPartsToTry = Math.min(numPartsToTry, partsScanned * limitScaleUpFactor)
+        }
+      }
 
-  override def requiredChildDistribution: List[Distribution] = AllTuples :: Nil
+      val p = partsScanned.until(math.min(partsScanned + numPartsToTry, totalParts).toInt)
+      val sc = sqlContext.sparkContext
+      val res = sc.runJob(childRDD,
+        (it: Iterator[InternalRow]) => Array[Int](it.size), p)
+
+      totalNum += res.map(_.head).sum
+      partsScanned += p.size
+
+      if (totalNum >= limit) {
+        // If we scan more rows than the limit number, we need to reduce that from scanned.
+        // We calculate how many rows need to be reduced for each partition,
+        // until all redunant rows are reduced.
+        var numToReduce = (totalNum - limit)
+        val reduceAmounts = new HashMap[Int, Int]()
+        val partitionsToReduce = p.zip(res.map(_.head)).foreach { case (part, size) =>
+          val toReduce = if (size > numToReduce) numToReduce else size
+          reduceAmounts += ((part, toReduce))
+          numToReduce -= toReduce
+        }
+        resultRDD = childRDD.mapPartitionsWithIndexInternal { case (index, iter) =>
+          if (index < partsScanned) {
```

Comment:
First, to limit a number less than the table has is corner case.

Second, no, it doesn't scan the table twice. It will compute all partitions, as before this change. Then, as the total number is less than the limit number, we won't drop any elements. So I can't get why you think it scans the table twice.